File: solutions/odd-even-linked-list/odd-even-linked-list.py
// @Title: Odd Even Linked List
// @Author: colinjxc
// @Date: 2022-02-03T14:26:17+08:00
// @URL: https://leetcode-cn.com/problems/odd-even-linked-list


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: ListNode) -> ListNode:
        if head == None:
            return head
        odd = head
        even = head.next
        even_head = even
        while even and even.next and odd.next:
            odd.next = odd.next.next
            even.next = even.next.next
            odd = odd.next
            even = even.next
        odd.next = even_head
        return head        

File: solutions/odd-even-linked-list/odd-even-linked-list.cpp
// @Title: Odd Even Linked List
// @Author: colinjxc
// @Date: 2022-02-03T02:09:50+08:00
// @URL: https://leetcode-cn.com/problems/odd-even-linked-list


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if (head == nullptr) {
            return head;
        }
        ListNode* odd = head;
        ListNode* evenHead = odd->next;
        ListNode* even = evenHead;
        while (even != nullptr && even->next != nullptr) {
            odd->next = odd->next->next;
            even->next = even->next->next;
            even = even->next;
            odd = odd->next;
        }
        odd->next = evenHead;
        return head;
    }
};
